cmd/server/main.go

```diff
@@ -1,53 +1,169 @@
+	runServer(config, store)
+	runIndexer(ctx, store)
+
+	runtime.Goexit()
+}
+
+func runIndexer(ctx context.Context, store *storage.Store) {
+	flowClient, err := grpc.NewBaseClient(
+		accessURL,
+		goGrpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
 	if err != nil {
-		fmt.Println("can't listen:", err)
+		panic(err)
 	}
-	defer listener.Close()
-	go server.ServeListener(listener)
-
-	requests := []string{
-		`{"jsonrpc":"2.0","id":1,"method":"eth_blockNumber","params": []}`,
-		`{"jsonrpc":"2.0","id":2,"method":"eth_chainId","params": []}`,
-		`{"jsonrpc":"2.0","id":3,"method":"eth_syncing","params": []}`,
-		`{"jsonrpc":"2.0","id":4,"method":"eth_getBalance","params": ["0x407d73d8a49eeb85d32cf465507dd71d507100c1"]}`,
-		`{"jsonrpc":"2.0","id":5,"method":"eth_getBlockTransactionCountByNumber","params": ["0x4E4ee"]}`,
+
+	latestBlockHeader, err := flowClient.GetLatestBlockHeader(ctx, true)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("Latest Block Height: %d\n", latestBlockHeader.Height)
+	fmt.Printf("Latest Block ID: %s\n", latestBlockHeader.ID)
+
+	data, errChan, initErr := flowClient.SubscribeEventsByBlockHeight(
+		ctx,
+		latestBlockHeader.Height,
+		flow.EventFilter{
+			Contracts: []string{"A.7e60df042a9c0868.FlowToken"},
+		},
+		grpc.WithHeartbeatInterval(1),
+	)
+	if initErr != nil {
+		log.Fatalf("could not subscribe to events: %v", initErr)
 	}
-	deadline := time.Now().Add(10 * time.Second)
 
-	for _, request := range requests {
-		conn, err := net.Dial("tcp", listener.Addr().String())
+	reconnect := func(height uint64) {
+		fmt.Printf("Reconnecting at block %d\n", height)
+
+		var err error
+		flowClient, err := grpc.NewBaseClient(
+			accessURL,
+			goGrpc.WithTransportCredentials(insecure.NewCredentials()),
+		)
 		if err != nil {
-			fmt.Println("can't dial:", err)
+			log.Fatalf("could not create flow client: %v", err)
 		}
 
-		conn.SetDeadline(deadline)
-		// Write the request, then half-close the connection so the server stops reading.
-		conn.Write([]byte(request))
-		conn.(*net.TCPConn).CloseWrite()
-		// Now try to get the response.
-		buf := make([]byte, 2000)
-		n, err := conn.Read(buf)
-		conn.Close()
+		data, errChan, initErr = flowClient.SubscribeEventsByBlockHeight(
+			ctx,
+			latestBlockHeader.Height,
+			flow.EventFilter{
+				Contracts: []string{"A.7e60df042a9c0868.FlowToken"},
+			},
+			grpc.WithHeartbeatInterval(1),
+		)
+		if initErr != nil {
+			log.Fatalf("could not subscribe to events: %v", initErr)
+		}
+	}
 
-		if err != nil {
-			fmt.Println("read error:", err)
+	// track the most recently seen block height. we will use this when reconnecting
+	// the first response should be for latestBlockHeader.Height
+	lastHeight := latestBlockHeader.Height - 1
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case response, ok := <-data:
+			if !ok {
+				if ctx.Err() != nil {
+					return // graceful shutdown
+				}
+				fmt.Println("subscription closed - reconnecting")
```

Comment:
I think the reconnect logic should have some timeout before reconnecting, otherwise a very vicious cycle can happen due to some issues in establishing connection. Also I would error log the connection failure so alarms can be set up in the future.